File: client/src/ot/BaseOtTests.ts
import * as assert from 'assert';
import { BaseOtAlice, BaseOtBob, decryptOtMessage } from './BaseOt';
import { NodeCrypto } from '../crypto/NodeCrypto';
import { SECP256K1_CURVE_ORDER, SECP256K1_FIELD_PRIME } from '../crypto/PointMath';

export interface BaseOtReport {
    allPassed: boolean;
}

export function runBaseOtUnitTests(): BaseOtReport {
    console.log('=============================================================');
    console.log(' Stage 5: Base Oblivious Transfer (COT.pdf Appendix A.3.1)');
    console.log('=============================================================\n');

    let allPassed = true;

    const checkTest = (name: string, condition: boolean, info: string = '') => {
        console.log(`  - [${condition ? 'PASS' : 'FAIL'}] ${name}${info ? ' -> ' + info : ''}`);
        if (!condition) allPassed = false;
    };

    // Deterministic test inputs
    const detA = Buffer.alloc(32); detA[31] = 7;
    const m0 = Buffer.alloc(32, 0x11);
    const m1 = Buffer.alloc(32, 0x22);

    // -----------------------------------------------------------------------
    // Test 1: Deterministic Base OT with choice c = 0
    // -----------------------------------------------------------------------
    console.log('--- Deterministic Base OT Test: choice c = 0 ---');
    {
        const detB0 = Buffer.alloc(32); detB0[31] = 11;
        const choice = 0;

        const alice = new BaseOtAlice();
        const A = alice.init(detA);
        checkTest('Alice init (a=7)', A.length === 33, `A=${A.toString('hex')}`);

        const bob = new BaseOtBob();
        const B0 = bob.receiveAAndComputeB(A, choice, detB0);
        checkTest('Bob receive A & compute B (b=11, c=0)', B0.length === 33, `B=${B0.toString('hex')}`);

        const { k0, k1, B_minus_A } = alice.receiveB(B0);
        checkTest('Alice receive B & derive keys', k0.length === 32 && k1.length === 32,
                  `k0=${k0.toString('hex')}, k1=${k1.toString('hex')}`);

        // Critical algebra checks
        const k0Match = k0.equals(bob.getKc());
        const k1Mismatch = !k1.equals(bob.getKc());
        checkTest('Algebra check: k0 == kc (c=0)', k0Match, `kc=${bob.getKc().toString('hex')}`);
        checkTest('Algebra check: k1 != kc (c=0)', k1Mismatch);

        // Round 3: Alice encrypts m0 and m1
        const { e0, e1 } = alice.encryptMessages(m0, m1);
        checkTest('Alice encrypt messages (m0, m1)', e0.length === 32 && e1.length === 32,
                  `e0=${e0.toString('hex')}, e1=${e1.toString('hex')}`);

        // Bob decrypts selected ciphertext e0 with kc
        const mc = bob.decryptSelected(e0, e1);
        const msgMatch = mc.equals(m0);
        checkTest('Bob decrypt selected: mc == m0 (c=0)', msgMatch, `Recovered: ${mc.toString('hex')}`);

        // Wrong-key decryption check
        const wrongDec = decryptOtMessage(e1, bob.getKc(), 1);
        const wrongKeyFails = !wrongDec.equals(m1);
        checkTest('Security check: unused ciphertext e1 cannot be decrypted with kc', wrongKeyFails);
    }
    console.log();

    // -----------------------------------------------------------------------
    // Test 2: Deterministic Base OT with choice c = 1
    // -----------------------------------------------------------------------
    console.log('--- Deterministic Base OT Test: choice c = 1 ---');
    {
        const detB1 = Buffer.alloc(32); detB1[31] = 13;
        const choice = 1;

        const alice = new BaseOtAlice();
        const A = alice.init(detA);
        checkTest('Alice init (a=7)', A.length === 33, `A=${A.toString('hex')}`);

        const bob = new BaseOtBob();
        const B1 = bob.receiveAAndComputeB(A, choice, detB1);
        checkTest('Bob receive A & compute B (b=13, c=1)', B1.length === 33, `B=${B1.toString('hex')}`);

        const { k0, k1, B_minus_A } = alice.receiveB(B1);
        checkTest('Alice receive B & derive keys', k0.length === 32 && k1.length === 32,
                  `k0=${k0.toString('hex')}, k1=${k1.toString('hex')}`);

        // Critical algebra checks
        const k1Match = k1.equals(bob.getKc());
        const k0Mismatch = !k0.equals(bob.getKc());
        checkTest('Algebra check: k1 == kc (c=1)', k1Match, `kc=${bob.getKc().toString('hex')}`);
        checkTest('Algebra check: k0 != kc (c=1)', k0Mismatch);

        // Round 3: Alice encrypts m0 and m1
        const { e0, e1 } = alice.encryptMessages(m0, m1);
        checkTest('Alice encrypt messages (m0, m1)', e0.length === 32 && e1.length === 32,
                  `e0=${e0.toString('hex')}, e1=${e1.toString('hex')}`);

        // Bob decrypts selected ciphertext e1 with kc
        const mc = bob.decryptSelected(e0, e1);
        const msgMatch = mc.equals(m1);
        checkTest('Bob decrypt selected: mc == m1 (c=1)', msgMatch, `Recovered: ${mc.toString('hex')}`);

        // Wrong-key decryption check
        const wrongDec = decryptOtMessage(e0, bob.getKc(), 0);
        const wrongKeyFails = !wrongDec.equals(m0);
        checkTest('Security check: unused ciphertext e0 cannot be decrypted with kc', wrongKeyFails);
    }
    console.log();

    // -----------------------------------------------------------------------
    // Test 3: Multiple Random In-Memory Protocol Iterations
    // -----------------------------------------------------------------------
    console.log('--- Random In-Memory Protocol Executions (5 runs) ---');
    {
        let allRandomOk = true;
        for (let run = 1; run <= 5; run++) {
            const randM0 = NodeCrypto.generateRandom32();
            const randM1 = NodeCrypto.generateRandom32();
            const randChoice = NodeCrypto.generateRandom32()[0] & 1;

            const alice = new BaseOtAlice();
            const A = alice.init();

            const bob = new BaseOtBob();
            const B = bob.receiveAAndComputeB(A, randChoice);

            alice.receiveB(B);

            const { e0, e1 } = alice.encryptMessages(randM0, randM1);
            const mc = bob.decryptSelected(e0, e1);

            const expectedM = (randChoice === 0) ? randM0 : randM1;
            if (!mc.equals(expectedM)) {
                allRandomOk = false;
            }
        }
        checkTest('5 independent random Base OT protocol runs', allRandomOk, 'All 5 recovered mc successfully');
    }
    console.log();

    // -----------------------------------------------------------------------
    // Test 4: Negative & Boundary Tests
    // -----------------------------------------------------------------------
    console.log('--- Base OT Negative / Boundary Tests ---');
    {
        const alice = new BaseOtAlice();
        const validA = alice.init();
        const bob = new BaseOtBob();

        // 1. Invalid choice bit (c = 2)
        let rejChoice2 = false;
        try {
            bob.receiveAAndComputeB(validA, 2);
        } catch {
            rejChoice2 = true;
        }
        checkTest('Reject invalid choice bit c = 2', rejChoice2);

        // 2. Invalid choice bit (c = 255)
        let rejChoice255 = false;
        try {
            bob.receiveAAndComputeB(validA, 255);
        } catch {
            rejChoice255 = true;
        }
        checkTest('Reject invalid choice bit c = 255', rejChoice255);

        // 3. Malformed point A with bad prefix 0x01
        const badA = Buffer.from(validA);
        badA[0] = 0x01;
        let rejBadA = false;
        try {
            bob.receiveAAndComputeB(badA, 0);
        } catch {
            rejBadA = true;
        }
        checkTest('Reject malformed point A with bad prefix 0x01', rejBadA);

        // 4. Malformed point B not on curve
        const badB = Buffer.alloc(33);
        badB[0] = 0x02;
        badB[32] = 0x05;
        let rejBadB = false;
        try {
            alice.receiveB(badB);
        } catch {
            rejBadB = true;
        }
        checkTest('Reject malformed point B not on curve', rejBadB);

        // 5. Scalar == 0 for Alice init
        const zeroScalar = Buffer.alloc(32);
        let rejZeroScalar = false;
        try {
            const badAlice = new BaseOtAlice();
            badAlice.init(zeroScalar);
        } catch {
            rejZeroScalar = true;
        }
        checkTest('Reject scalar == 0 for Alice init', rejZeroScalar);

        // 6. Scalar == n for Alice init
        const nHex = SECP256K1_CURVE_ORDER.toString(16).padStart(64, '0');
        const nScalar = Buffer.from(nHex, 'hex');
        let rejNScalar = false;
        try {
            const badAlice = new BaseOtAlice();
            badAlice.init(nScalar);
        } catch {
            rejNScalar = true;
        }
        checkTest('Reject scalar == n for Alice init', rejNScalar);

        // 7. Scalar == p - 1 for Alice init (since p - 1 > n)
        const pMinus1Hex = (SECP256K1_FIELD_PRIME - 1n).toString(16).padStart(64, '0');
        const pMinus1Scalar = Buffer.from(pMinus1Hex, 'hex');
        let rejPMinus1 = false;
        try {
            const badAlice = new BaseOtAlice();
            badAlice.init(pMinus1Scalar);
        } catch {
            rejPMinus1 = true;
        }
        checkTest('Reject scalar == p - 1 for Alice init', rejPMinus1);

        // 8. Wrong length point
        let rejWrongLenPoint = false;
        try {
            bob.receiveAAndComputeB(Buffer.alloc(32), 0);
        } catch {
            rejWrongLenPoint = true;
        }
        checkTest('Reject wrong-length point (32 bytes)', rejWrongLenPoint);

        // 9. Wrong length scalar
        let rejWrongLenScalar = false;
        try {
            const badAlice = new BaseOtAlice();
            badAlice.init(Buffer.alloc(31));
        } catch {
            rejWrongLenScalar = true;
        }
        checkTest('Reject wrong-length scalar (31 bytes)', rejWrongLenScalar);
    }

    console.log('\n=============================================================');
    console.log(` Stage 5 TypeScript Base OT Test Result: ${allPassed ? 'ALL PASSED' : 'FAILED'}`);
    console.log('=============================================================\n');

    return { allPassed };
}

File: server/src/ot/BaseOtTests.cpp
#include "ot/BaseOtTests.hpp"
#include "ot/BaseOt.hpp"
#include "crypto/TrezorCrypto.hpp"

#include <iostream>
#include <iomanip>
#include <cassert>
#include <cstring>
#include <vector>

namespace cot::ot {

bool run_base_ot_unit_tests() {
    std::cout << "=========================================================\n";
    std::cout << " Stage 5: Base Oblivious Transfer (COT.pdf Appendix A.3.1)\n";
    std::cout << "=========================================================\n\n";

    bool all_passed = true;

    auto check_test = [&](const std::string& name, bool condition, const std::string& info = "") {
        std::cout << "  - [" << (condition ? "PASS" : "FAIL") << "] " << name;
        if (!info.empty()) {
            std::cout << " -> " << info;
        }
        std::cout << "\n";
        if (!condition) all_passed = false;
    };

    // Deterministic test setup
    uint8_t det_a[32] = {0}; det_a[31] = 7;
    uint8_t m0[32]; std::memset(m0, 0x11, 32);
    uint8_t m1[32]; std::memset(m1, 0x22, 32);

    // -----------------------------------------------------------------------
    // Test 1: Deterministic Base OT with choice c = 0
    // -----------------------------------------------------------------------
    std::cout << "--- Deterministic Base OT Test: choice c = 0 ---\n";
    {
        uint8_t det_b0[32] = {0}; det_b0[31] = 11;
        uint8_t choice = 0;

        BaseOtAlice alice;
        bool alice_ok = alice.init(det_a);
        check_test("Alice init (a=7)", alice_ok,
                   "A=" + crypto::to_hex(alice.get_A(), 33));

        BaseOtBob bob;
        uint8_t B[33] = {0};
        bool bob_ok = bob.receive_A_and_compute_B(alice.get_A(), choice, B, det_b0);
        check_test("Bob receive A & compute B (b=11, c=0)", bob_ok,
                   "B=" + crypto::to_hex(B, 33));

        bool alice_rx_b = alice.receive_B(B);
        check_test("Alice receive B & derive keys", alice_rx_b,
                   "k0=" + crypto::to_hex(alice.get_k0(), 32) +
                   ", k1=" + crypto::to_hex(alice.get_k1(), 32));

        // Critical algebra checks:
        // When c = 0: k0 must match kc, and k1 must not match kc
        bool k0_match = (std::memcmp(alice.get_k0(), bob.get_kc(), 32) == 0);
        bool k1_mismatch = (std::memcmp(alice.get_k1(), bob.get_kc(), 32) != 0);
        check_test("Algebra check: k0 == kc (c=0)", k0_match,
                   "kc=" + crypto::to_hex(bob.get_kc(), 32));
        check_test("Algebra check: k1 != kc (c=0)", k1_mismatch);

        // Round 3: Alice encrypts m0 and m1
        uint8_t e0[32] = {0}, e1[32] = {0};
        bool enc_ok = alice.encrypt_messages(m0, m1, e0, e1);
        check_test("Alice encrypt messages (m0, m1)", enc_ok,
                   "e0=" + crypto::to_hex(e0, 32) +
                   ", e1=" + crypto::to_hex(e1, 32));

        // Bob decrypts selected ciphertext e0 with kc
        uint8_t mc[32] = {0};
        bool dec_ok = bob.decrypt_selected(e0, e1, mc);
        bool msg_match = (std::memcmp(mc, m0, 32) == 0);
        check_test("Bob decrypt selected: mc == m0 (c=0)", dec_ok && msg_match,
                   "Recovered: " + crypto::to_hex(mc, 32));

        // Wrong-key decryption check: attempt to decrypt e1 using key kc (which is k0 != k1)
        uint8_t wrong_dec[32] = {0};
        decrypt_ot_message(e1, bob.get_kc(), 1, wrong_dec);
        bool wrong_key_fails = (std::memcmp(wrong_dec, m1, 32) != 0);
        check_test("Security check: unused ciphertext e1 cannot be decrypted with kc", wrong_key_fails);
    }
    std::cout << "\n";

    // -----------------------------------------------------------------------
    // Test 2: Deterministic Base OT with choice c = 1
    // -----------------------------------------------------------------------
    std::cout << "--- Deterministic Base OT Test: choice c = 1 ---\n";
    {
        uint8_t det_b1[32] = {0}; det_b1[31] = 13;
        uint8_t choice = 1;

        BaseOtAlice alice;
        bool alice_ok = alice.init(det_a);
        check_test("Alice init (a=7)", alice_ok,
                   "A=" + crypto::to_hex(alice.get_A(), 33));

        BaseOtBob bob;
        uint8_t B[33] = {0};
        bool bob_ok = bob.receive_A_and_compute_B(alice.get_A(), choice, B, det_b1);
        check_test("Bob receive A & compute B (b=13, c=1)", bob_ok,
                   "B=" + crypto::to_hex(B, 33));

        bool alice_rx_b = alice.receive_B(B);
        check_test("Alice receive B & derive keys", alice_rx_b,
                   "k0=" + crypto::to_hex(alice.get_k0(), 32) +
                   ", k1=" + crypto::to_hex(alice.get_k1(), 32));

        // Critical algebra checks:
        // When c = 1: k1 must match kc, and k0 must not match kc
        bool k1_match = (std::memcmp(alice.get_k1(), bob.get_kc(), 32) == 0);
        bool k0_mismatch = (std::memcmp(alice.get_k0(), bob.get_kc(), 32) != 0);
        check_test("Algebra check: k1 == kc (c=1)", k1_match,
                   "kc=" + crypto::to_hex(bob.get_kc(), 32));
        check_test("Algebra check: k0 != kc (c=1)", k0_mismatch);

        // Round 3: Alice encrypts m0 and m1
        uint8_t e0[32] = {0}, e1[32] = {0};
        bool enc_ok = alice.encrypt_messages(m0, m1, e0, e1);
        check_test("Alice encrypt messages (m0, m1)", enc_ok,
                   "e0=" + crypto::to_hex(e0, 32) +
                   ", e1=" + crypto::to_hex(e1, 32));

        // Bob decrypts selected ciphertext e1 with kc
        uint8_t mc[32] = {0};
        bool dec_ok = bob.decrypt_selected(e0, e1, mc);
        bool msg_match = (std::memcmp(mc, m1, 32) == 0);
        check_test("Bob decrypt selected: mc == m1 (c=1)", dec_ok && msg_match,
                   "Recovered: " + crypto::to_hex(mc, 32));

        // Wrong-key decryption check: attempt to decrypt e0 using key kc (which is k1 != k0)
        uint8_t wrong_dec[32] = {0};
        decrypt_ot_message(e0, bob.get_kc(), 0, wrong_dec);
        bool wrong_key_fails = (std::memcmp(wrong_dec, m0, 32) != 0);
        check_test("Security check: unused ciphertext e0 cannot be decrypted with kc", wrong_key_fails);
    }
    std::cout << "\n";

    // -----------------------------------------------------------------------
    // Test 3: Multiple Random In-Memory Protocol Iterations
    // -----------------------------------------------------------------------
    std::cout << "--- Random In-Memory Protocol Executions (5 runs) ---\n";
    {
        bool all_random_ok = true;
        for (int run = 1; run <= 5; ++run) {
            uint8_t rand_m0[32], rand_m1[32];
            crypto::generate_random_32(rand_m0);
            crypto::generate_random_32(rand_m1);

            uint8_t rand_buf[32];
            crypto::generate_random_32(rand_buf);
            uint8_t choice_bit = rand_buf[0] & 1;

            BaseOtAlice alice;
            alice.init();

            BaseOtBob bob;
            uint8_t B[33];
            bob.receive_A_and_compute_B(alice.get_A(), choice_bit, B);

            alice.receive_B(B);

            uint8_t e0[32], e1[32];
            alice.encrypt_messages(rand_m0, rand_m1, e0, e1);

            uint8_t mc[32];
            bool dec_ok = bob.decrypt_selected(e0, e1, mc);

            const uint8_t* expected_m = (choice_bit == 0) ? rand_m0 : rand_m1;
            if (!dec_ok || std::memcmp(mc, expected_m, 32) != 0) {
                all_random_ok = false;
            }
        }
        check_test("5 independent random Base OT protocol runs", all_random_ok, "All 5 recovered mc successfully");
    }
    std::cout << "\n";

    // -----------------------------------------------------------------------
    // Test 4: Negative & Boundary Tests
    // -----------------------------------------------------------------------
    std::cout << "--- Base OT Negative / Boundary Tests ---\n";
    {
        BaseOtBob bob;
        uint8_t valid_A[33];
        BaseOtAlice alice;
        alice.init();
        std::memcpy(valid_A, alice.get_A(), 33);

        uint8_t dummy_B[33];

        // 1. Invalid choice bit (choice = 2)
        bool rej_choice_2 = !bob.receive_A_and_compute_B(valid_A, 2, dummy_B);
        check_test("Reject invalid choice bit c = 2", rej_choice_2);

        // 2. Invalid choice bit (choice = 255)
        bool rej_choice_255 = !bob.receive_A_and_compute_B(valid_A, 255, dummy_B);
        check_test("Reject invalid choice bit c = 255", rej_choice_255);

        // 3. Malformed point A (bad prefix 0x01)
        uint8_t bad_A[33];
        std::memcpy(bad_A, valid_A, 33);
        bad_A[0] = 0x01;
        bool rej_bad_A = !bob.receive_A_and_compute_B(bad_A, 0, dummy_B);
        check_test("Reject malformed point A with bad prefix 0x01", rej_bad_A);

        // 4. Malformed point B passed to Alice
        uint8_t bad_B[33] = {0};
        bad_B[0] = 0x02;
        bad_B[32] = 0x05; // Non-quadratic residue, not on curve
        bool rej_bad_B = !alice.receive_B(bad_B);
        check_test("Reject malformed point B not on curve", rej_bad_B);

        // 5. Invalid scalar for Alice init (scalar = 0)
        uint8_t zero_scalar[32] = {0};
        BaseOtAlice bad_alice;
        bool rej_zero_scalar = !bad_alice.init(zero_scalar);
        check_test("Reject scalar == 0 for Alice init", rej_zero_scalar);

        // 6. Invalid scalar for Alice init (scalar = n)
        bool rej_n_scalar = !bad_alice.init(crypto::SECP256K1_CURVE_ORDER_BE.data());
        check_test("Reject scalar == n for Alice init", rej_n_scalar);
    }

    std::cout << "\n=========================================================\n";
    std::cout << " Stage 5 C++ Base OT Test Result: " << (all_passed ? "ALL PASSED" : "FAILED") << "\n";
    std::cout << "=========================================================\n\n";

    return all_passed;
}

} // namespace cot::ot
